### PR TITLE
LC-338 application-example.conf documentation

### PR DESCRIPTION
**Pull request by abijitran**
Reorganized application-example.conf and went through all mentions of config properties for Lucille items other than stages and plug-ins. There is definitely more to be added from reviewers and some unknown properties that need to be questioned / resolved.

### CODE REVIEW

File: application-example.conf [hunk 1]
```diff
@@ -1,133 +1,192 @@
...
+
+    maxRequestSize: 250000000
+
+    securityProtocol: "SSL"
+
+    consumerPropertyFile: ".../consumer-conf/consumer.properties"
+
+    producerPropertyFile: ".../consumer-conf/producer.properties"
+
+    documentDeserializer: # TODO
```
Comment (abijitran) [inline application-example.conf:151]: Am unsure what these items are doing, I see that KafkaUtils.java is placing them in a properties map that is then fed into the AdminUser creation, but not sure what a good example for these properties might be.

**Review by abijitran — COMMENTED** [review]
Self-reviewing to pose a question

File: application-example.conf [hunk 2]
```diff
@@ -1,133 +1,192 @@
...
   }
+## ----------------------------------------------------------------------------------------------------------------------------------
 
-  elastic {
-    host: "localhost",
-    port: "9200"
-    sendEnabled: ${?ELASTIC_SEND_ENABLED}
+
+## Misc: -----------------------------------------------------------------------------------------------------------------
+  # TODO: Do we still need this dev block and relevant features?
```
Comment (meghanboyd) [inline application-example.conf:165]: Voted in Lucille planning to remove

File: application-example.conf (hunk 1 above)
Comment (abijitran) [reply application-example.conf:151]: Follow up: Might need to add tests to cover a lot of these properties, as they currently don't exist

File: lucille-core/src/main/java/com/kmwllc/lucille/core/ConfigUtils.java [hunk 3]
```diff
@@ -25,9 +25,9 @@ public static Config loadConfig() throws ConfigException {
 
     log.info("pipeline.env system property = " + envName);
     // Override default settings in application.conf with settings for the named environment.
-    // For example, if application.conf contains x=1 and dev.x=2, and if the the
+    // For example, if application.conf contains x=1 and dev.x=2, and if the
     // pipeline.env system property is set to "dev", then accessing "x" will return 2
-    // even though the "dev" prefix was not included
+    // even though the "dev" prefix was not provided when accessing "x"
```
Comment (rseitz) [inline lucille-core/src/main/java/com/kmwllc/lucille/core/ConfigUtils.java:30]: this entire method is obsolete now so we should remove it

File: application-example.conf [hunk 4]
```diff
@@ -89,45 +137,35 @@ kafka {
...
 
-  # maximum number of times across all workers that an attempt should be made to process any given document;
-  # when this property is not provided, retries will not be tracked and no limit will be imposed;
-  # this property requires that zookeeper is available at the specified zookeeper.connectString
-  #maxRetries: 2
+  producerPropertyFile: ".../consumer-conf/producer.properties"
 
-  # number of worker threads to start for each pipeline when running lucille in local / single-JVM mode
-  threads: 2
+  documentDeserializer: # TODO
```
Comment (rseitz) [inline application-example.conf:146]: can use "com.kmwllc.lucille.message.KafkaDocumentDeserializer"
try to mention that this is the default and doesn't need to be set explicitly unless someone wants to use a custom implementation of Deserializer<Document>

Deserializer<Document> is an implementation of org.apache.kafka.common.serialization.Deserializer that can convert a message on a Kafka topic into a Lucille Document

File: application-example.conf [hunk 5]
```diff
@@ -89,45 +137,35 @@ kafka {
...
 
-  # maximum time to spend processing a message, after which the worker will assume a problem and shut down
-  # NOTE: not supported while com.kmwllc.lucille.core.Worker and PipelineWorker are being reconciled
-  #maxProcessingSecs: 600 # 10 minutes
-}
+  documentSerializer: # TODO
```
Comment (rseitz) [inline application-example.conf:148]: can use "com.kmwllc.lucille.message.KafkaDocumentSerializer"
try to mention that this is the default and doesn't need to be set explicitly unless someone wants to use a custom implementation of Serializer<Document>

Serializer<Document> is an implementation of org.apache.kafka.common.serialization.Serializer that can convert a Lucille document into a byte[] that can be written as a message on a Kafka topic

File: application-example.conf [hunk 6]
```diff
@@ -89,45 +137,35 @@ kafka {
...
 
-publisher {
+  events: # TODO
```
Comment (rseitz) [inline application-example.conf:150]: true is the default
can set to false to indicate that "events" (indicating whether a document succeeded or failed indexing) should not be sent as Kafka message
a setting of false would be used in a context where Lucille is running without a Runner/Publisher that would receive events and determine when the run is complete; in this context, there might be a WorkerIndexer running that reads documents directly from a given Kafka topic and sends them to the search engine without tracking successes or failures

File: application-example.conf [hunk 7]
```diff
@@ -89,45 +137,35 @@ kafka {
...
-  # this setting controls
-  #     1) maximum size of the queue of published documents waiting to be processed
-  #     2) maximum size of the queue of completed documents waiting to be indexed
-  # e.g. if it is set to 10 then each each queue can contain up to 10 documents for a total of 20
-  #
-  # attempts to publish beyond this capacity will cause the publisher to block
-  # this setting applies only when running Lucille in local / single-JVM mode using a LocalMessageManager
-  # this setting defaults to 100
-  queueCapacity: 100
+  sourceTopic: # TODO
```
Comment (rseitz) [inline application-example.conf:152]: this is the name of the kafka topic that contains Documents generated by the Publisher that are waiting to be picked up by a Worker and sent through the pipeline. Typically, lucille generates this topic "behind the scenes" and names it as <pipeline_name>_source
However, there are cases where Lucille runs without a Runner/Publisher and doesn't create the sourceTopic; rather an existing Kafka topic should be used as sourceTopic and a lucille Worker or WorkerIndexer should read Documents from that existing topic. In that scenario, a user would want to specify kafka.sourceTopic in the config and make sure it's set to whatever topic contains the documents that they want to process

File: application-example.conf (hunk 1 above)
Comment (abijitran) [reply application-example.conf]: Resolved

File: application-example.conf (hunk 6 above)
Comment (abijitran) [reply application-example.conf]: Thanks! Addressed in new commit

File: application-example.conf (hunk 7 above)
Comment (abijitran) [reply application-example.conf]: Thanks! Addressed in new commit

File: application-example.conf [hunk 8]
```diff
@@ -1,80 +1,130 @@
...
+  # maximum number of times across all workers that an attempt should be made to process any given document;
+  # when this property is not provided, retries will not be tracked and no limit will be imposed;
+  # this property requires that zookeeper is available at the specified zookeeper.connectString
+  maxRetries: 2
+
+  # number of worker threads to start for each pipeline when running lucille in local / single-JVM mode
+  threads: 2
+
+  # maximum time to spend processing a message, after which the worker will assume a problem and shut down
+  # NOTE: not supported while com.kmwllc.lucille.core.Worker and PipelineWorker are being reconciled
```
Comment (rseitz) [inline application-example.conf:119]: this "NOTE:" is obsolete & should be removed -- line 119 only

but on line 118 where it says "assume a problem and shut down," need to clarify that shut down only happens if worker.exitOnTimeout=true

as discussed, also want to mention worker.heartbeat and worker.period

Comment (abijitran) [reply application-example.conf]: Addressed in new commit. However, if you could take another look at the comments and explanation I provided, I just want to make sure that it is accurate

**Review by rseitz — APPROVED** [review]

**Review by meghanboyd — APPROVED** [review]

**Pull request merged**